cmd/terramate/e2etests/safeguard_test.go

```diff
@@ -0,0 +1,292 @@
+// Copyright 2022 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2etest
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/mineiros-io/terramate/cmd/terramate/cli"
+	"github.com/mineiros-io/terramate/test"
+	"github.com/mineiros-io/terramate/test/sandbox"
+)
+
+func TestSafeguardNotRequiredInSomeCommands(t *testing.T) {
+	t.Parallel()
+
+	// Regression test to guarantee that all git checks
+	// are disabled and no git operation will be performed for certain commands.
+	// Some people like to get some coding done on airplanes :-)
+	s := sandbox.New(t)
+	git := s.Git()
+	git.SetRemoteURL("origin", "http://non-existant/terramate.git")
+
+	s.CreateStack("stack")
+
+	cli := newCLI(t, s.RootDir())
+
+	cmds := []string{
+		"experimental metadata",
+		"experimental globals",
+		"experimental run-order",
+		"experimental run-graph",
+		"create stack-2",
+		"generate",
+		"list",
+	}
+	for _, cmd := range cmds {
+		t.Run(cmd, func(t *testing.T) {
+			args := strings.Split(cmd, " ")
+			assertRunResult(t, cli.run(args...), runExpected{
+				IgnoreStdout: true,
+			})
+		})
+	}
+}
+
+func TestSafeguardFailsOnRunIfRemoteMainIsOutdated(t *testing.T) {
+	t.Parallel()
+
+	s := sandbox.New(t)
+
+	stack := s.CreateStack("stack-1")
+	mainTfFile := stack.CreateFile("main.tf", "# no code")
+
+	ts := newCLI(t, s.RootDir())
+
+	git := s.Git()
+
+	git.Add(".")
+	git.Commit("all")
+
+	setupLocalMainBranchBehindOriginMain(git, func() {
+		stack.CreateFile("tempfile", "any content")
+	})
+
+	testrun := func() {
+		wantRes := runExpected{
+			Status:      1,
+			StderrRegex: string(cli.ErrCurrentHeadIsOutOfSync),
+		}
+
+		assertRunResult(t, ts.run(
+			"run",
+			testHelperBin,
+			"cat",
+			mainTfFile.HostPath(),
+		), wantRes)
+
+		assertRunResult(t, ts.run(
+			"run",
+			"--changed",
+			testHelperBin,
+			"cat",
+			mainTfFile.HostPath(),
+		), wantRes)
+	}
+
+	testrun()
+
+	git.CheckoutNew("branch")
+
+	// we create two commits so we can also test from a DETACHED HEAD.
+	stack.CreateFile("tempfile2", "any content")
+	git.CommitAll("add tempfile2")
+
+	stack.CreateFile("tempfile3", "any content")
+	git.CommitAll("add tempfile3")
+
+	testrun()
+
+	git.Checkout("HEAD^1")
+
+	testrun()
+}
+
+func TestSafeguardDisableGitCheckRemoteFromArgs(t *testing.T) {
```

Comment:
```suggestion
func TestSafeguardDisableGitCheckRemote(t *testing.T) {
```

We have subtests to disable in different ways, not just args.